orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/DetermineRollbackCandidatesTask.java

```diff
@@ -281,6 +234,172 @@ public TaskResult execute(@Nonnull StageExecution stage) {
+    ImmutableMap.Builder<Object, Object> imageToRestore =
+        ImmutableMap.builder()
+            .put("region", region)
+            .put("image", rollbackDetails.imageName)
+            .put("rollbackMethod", rollbackDetails.rollbackType.toString());
+
+    if (rollbackDetails.buildNumber != null) {
+      imageToRestore.put("buildNumber", rollbackDetails.buildNumber);
+    }
+
+    return imageToRestore.build();
+  }
+
+  @Nonnull
+  private Map<String, Object> getRollbackContext(
+      @Nullable Integer targetHealthyRollbackPercentage,
+      ServerGroup serverGroupToRollBack,
+      RollbackDetails rollbackDetails) {
+    Map<String, Object> rollbackContext = new HashMap<>(rollbackDetails.rollbackContext);
+    rollbackContext.put(
+        "targetHealthyRollbackPercentage",
+        determineTargetHealthyRollbackPercentage(
+            serverGroupToRollBack.capacity, targetHealthyRollbackPercentage));
+    return rollbackContext;
+  }
+
+  @Nonnull
+  private DetermineRollbackCandidatesTask.RollbackDetails getDetailsUsingPreviousServerGroups(
+      List<ServerGroup> candidateServerGroupsInRegion,
+      ServerGroup serverGroupToRollBack,
+      String cluster,
+      String region) {
+
+    logger.info(
+        "Looking for rollback candidates in cluster {}, region {} based on previous server groups. ",
+        cluster,
+        region);
+    return fetchRollbackDetails(serverGroupToRollBack, candidateServerGroupsInRegion);
+  }
+
+  /** Check for rollback candidates based on entity tags */
+  @Nonnull
+  private DetermineRollbackCandidatesTask.RollbackDetails getDetailsUsingEntityTags(
+      List<ServerGroup> candidateServerGroupsInRegion,
+      ServerGroup serverGroupToRollBack,
+      ImageDetails imageDetails,
+      String cluster,
+      String region) {
+
+    logger.info(
+        "Looking for rollback candidates in cluster {}, region {} based on entity tags. ",
+        cluster,
+        region);
+
+    return fetchRollbackDetails(imageDetails, serverGroupToRollBack, candidateServerGroupsInRegion);
+  }
+
+  private boolean shouldOnlyConsiderEnabledServerGroups(
+      Map<String, Object> additionalRollbackContext) {
+    return Optional.ofNullable(additionalRollbackContext)
+        .map(a -> (Boolean) a.get("onlyEnabledServerGroups"))
+        .orElse(false);
+  }
+
+  /**
+   * Retrieve info about the server group and use it to populate a Moniker object
+   *
+   * @return true on success, false on failure
+   */
+  private AtomicReference<Moniker> populateMonikerWithServerGroupInfo(
+      Moniker m, String credentials, String region, String serverGroupName) {
+    AtomicReference<Moniker> moniker = new AtomicReference<>(m);
+    if (moniker.get() == null && serverGroupName != null) {
+      try {
+        Map<String, Object> serverGroup =
+            retrySupport.retry(
+                () -> fetchServerGroup(credentials, region, serverGroupName), 5, 1000, false);
```

Comment:
later: `RetrySupport` usage should be removed in favor of `Retry` from resilience4j.